### ISSUE TITLE
How Do I? : Embed dtale in existing flask application without having a separate port open 

### ISSUE THREAD

**Issue by bigbprp** [man-group/dtale#282]
I want to embed dlate in an existing flask app. what all options do I have? 

my code looks like as follows:

`from flask import Flask, render_template, request, jsonify, json, url_for, redirect
import pandas as pd
import dtale
import dtale.app as dtale_app
import dtale  `

app = Flask(__name__)
@app.route("/")
def hello():
    dataf = pd.DataFrame({'num_legs': [2, 4, 8, 0],
                          'num_wings': [2, 0, 0, 0],
                          'num_specimen_seen': [10, 2, 1, 8]},
                         index=['falcon', 'dog', 'spider', 'fish'])`

    dd = dtale.show(dataf, host='localhost',
                    ignore_duplicate=True)
    dd._main_url
    dd
    return render_template('myanalysis.html', qr=dd._main_url)

`
if __name__ == '__main__':
    app.run()
`
not sure why but using app_root didn't help much. I am not sure what am I missing? The whole object is not to open a separate port.

**Comment by aschonfeld** [comment]
@bigbprp a similar question was asked a while back, please take a look at this [thread](https://github.com/man-group/dtale/issues/256).  If this gets you where you need to go let me know and I'll close this otherwise if you're still having issues let me know and I'll dig in a little deeper.

**Comment by bigbprp** [comment]
I did some basic changes & I could be limited by my knowledge of implementing Flask based application. 
code shared in the above thread didn't work for me. 

code goes as follows:


from flask import redirect
from dtale.app import DtaleFlask
from dtale.views import startup
import pandas as pd
import dtale

class MyDtaleFlask(DtaleFlask):
    def __init__(self, *args, **kwargs):  
        super(MyDtaleFlask, self).__init__(*args, **kwargs)


app = MyDtaleFlask(__name__)

@app.route("/")
def home():
   
    dataf = pd.DataFrame({'num_legs': [2, 4, 8, 0],
                          'num_wings': [2, 0, 0, 0],
                          'num_specimen_seen': [10, 2, 1, 8]},
                         index=['falcon', 'dog', 'spider', 'fish'])

    pivot_instance = startup('',
                             data=dataf, ignore_duplicate=True)
    return str(dtale.get_instance(pivot_instance._data_id)._main_url) + " pivot instance details : " + str(pivot_instance._data_id)
    --return redirect(dtale.get_instance(pivot_instance._data_id)._main_url, code=302)

if __name__ == '__main__':
    app.run(host='localhost', port=5001)

Two issues:
1. def __init__(self, import_name, *args, **kwargs):  Here `import_name` gave an error so, I had to remove the same from function. 
2. return redirect(dtale.get_instance(pivot_instance._data_id)._main_url, code=302) gave me following output. so I am not sure how to proceed now. 
2.ahttp://None/dtale/main/1 pivot instance details : 1

Kindly, guide on closure.

**Comment by bigbprp** [comment]
Let me call out my ask again. This is simple. 
My main application URL: http://myhost:port/Myapplication
I want dtale to be available at : http://myhost:port/dtale/main/1  ( I don't want to open a different port)

**Comment by aschonfeld** [comment]
@bigprp Alrighty, after a little hacking I got it to work.  There's still some more work to make it a little interactive but here's the code: 
```
import pandas as pd

from flask import redirect

from dtale.app import build_app, initialize_process_props
from dtale.views import startup
from dtale.utils import build_url


HOST = "0.0.0.0"
PORT = 8080

initialize_process_props(HOST, PORT)
app_url = build_url(HOST, str(PORT))
app = build_app(app_url, reaper_on=False)


@app.route("/create-df")
def create_df():
    df = pd.DataFrame(dict(a=[1, 2, 3], b=[4, 5, 6]))
    instance = startup("", data=df, ignore_duplicate=True)

    return redirect(f"/dtale/main/{instance._data_id}", code=302)


if __name__ == '__main__':
    app.run(host=HOST, port=PORT)
```

And here is a little demo for proof: https://youtu.be/wG6HC8HFVu0

Good luck!

**Comment by bigbprp** [comment]
@aschonfeld , Thank you very much for your guidance. Maybe I am missing something. After executing code as it is I am getting the following error:

This could be something very trivial that I am missing. Apologies for the same... 

  File "/home/appl1/anaconda3/lib/python3.8/site-packages/flask/app.py", line 2447, in wsgi_app
    response = self.full_dispatch_request()
  File "/home/appl1/anaconda3/lib/python3.8/site-packages/flask/app.py", line 1952, in full_dispatch_request
    rv = self.handle_user_exception(e)
  File "/home/appl1/anaconda3/lib/python3.8/site-packages/flask/app.py", line 1821, in handle_user_exception
    reraise(exc_type, exc_value, tb)
  File "/home/appl1/anaconda3/lib/python3.8/site-packages/flask/_compat.py", line 39, in reraise
    raise value
  File "/home/appl1/anaconda3/lib/python3.8/site-packages/flask/app.py", line 1950, in full_dispatch_request
    rv = self.dispatch_request()
  File "/home/appl1/anaconda3/lib/python3.8/site-packages/flask/app.py", line 1936, in dispatch_request
    return self.view_functions[rule.endpoint](**req.view_args)
  File "/home/appl1/anaconda3/lib/python3.8/site-packages/dtale/app.py", line 258, in root
    return redirect("/dtale/main/{}".format(head_data_id()))
  File "/home/appl1/anaconda3/lib/python3.8/site-packages/dtale/views.py", line 87, in head_data_id
    raise Exception("No data associated with this D-Tale session")
Exception: No data associated with this D-Tale session

Let me try again & revert with closure on this thread.

**Comment by aschonfeld** [comment]
Yep, sorry I probably should have mentioned that currently in order to get the app to show up you can't hit the default endpoint of `0.0.0.0:8080` because that is mapped to the D-Tale main endpoint which requires so data to have been loaded in order for it to work.

That being said I hope to turn this exception off in my next release because of the fact I've given users the ability to load data via the "Upload CSV/TSV" button.

In order to work around it in the code I sent you there is two options:
1) jump directly to `0.0.0.0:8080/create-df` which will trigger data to be loaded to D-Tale and thus eliminating that exception from being thrown.
2) you can try overriding the `/` endpoint so that it won't go directly to D-Tale before you've loaded data.  Currently this is the code in dtale.app that configures that endpoint:
```
@app.route("/")
    @app.route("/dtale")
    def root():
        """
        :class:`flask:flask.Flask` routes which redirect to dtale/main

        :return: 302 - flask.redirect('/dtale/main')
        """
        return redirect("/dtale/main/{}".format(head_data_id()))
```
So until I get that change in here is some updated code with a workaround:
```
import pandas as pd

from flask import redirect

from dtale.app import build_app, initialize_process_props
from dtale.views import startup
from dtale.utils import build_url

if __name__ == '__main__':
    HOST = "0.0.0.0"
    PORT = 8080

    initialize_process_props(HOST, PORT)
    app_url = build_url(HOST, str(PORT))
    app = build_app(app_url, reaper_on=False)

    @app.route("/create-df")
    def create_df():
        df = pd.DataFrame(dict(a=[1, 2, 3], b=[4, 5, 6]))
        instance = startup("", data=df, ignore_duplicate=True)

        return redirect(f"/dtale/main/{instance._data_id}", code=302)

    @app.route("/")
    def hello_world():
        return 'Hi there, load data using <a href="/create-df">create-df</a>'

    app.url_map._rules = [r for r in app.url_map._rules if not (r.rule == '/' and r.endpoint == 'root')]
    app.url_map._remap = True
    app.url_map.update()

    app.run(host=HOST, port=PORT)
```

**Comment by bigbprp** [comment]
@aschonfeld , Thank you very much. This resolved the issue. Your recommended solution worked.

**Comment by aschonfeld** [comment]
@bigbprp Just wanted to give you a heads up that I just released v1.18.0 which includes better handling for embedding D-Tale within another Flask app.  Here is what that code snippet I sent earlier will look like in the new version:
```
from flask import redirect

from dtale.app import build_app
from dtale.views import startup

if __name__ == '__main__':
    app = build_app(reaper_on=False)

    @app.route("/create-df")
    def create_df():
        df = pd.DataFrame(dict(a=[1, 2, 3], b=[4, 5, 6]))
        instance = startup("", data=df, ignore_duplicate=True)

        return redirect(f"/dtale/main/{instance._data_id}", code=302)

    @app.route("/")
    def hello_world():
        return 'Hi there, load data using <a href="/create-df">create-df</a>'

    app.run(host="0.0.0.0", port=8080)
```

In addition to these updates I've also given users the ability to start D-Tale without requiring data to be loaded up front.  So now you can start D-Tale in two ways:
1) `dtale.show()`
2) From the command line: `> dtale --port 8080`

Once you've started it without data you will be redirected to a screen where you can load either a CSV or TSV file.

**Comment by ghawes85** [comment]
Hey @aschonfeld Thanks for all the work you are putting into dtale. It is really awesome! One minor thing I am struggling with when adding it to my Flask application, is I am getting the data import popup (dtale/popup/upload) right away on startup. This would be great to prompt the user, but my use-case for dtale is it would be a part of the overall application, and the user may not be entering or looking at data right away. Is there a way to suppress the popup? Or call it when navigating to a certain route depending on if data is available or not already?

**Comment by aschonfeld** [comment]
So the way to override that would be to implement your own flask route against the base path:
```
@app.route('/'):
def my_root():
  return send_the_app_to_wherever_you_want_to_go...
```
This will override the root path mapping that D-Tale has and should eliminate your problem.  Make sure you have the latest version too because I added a fix for overriding routes that will work with gunicorn as well.

**Comment by sk1802** [comment]
Hi @aschonfeld ,Thank you for all the work you are doing, I had a minor doubt reagrding the integration of dtale. On my flask app there is a template where 200+ datasets are shown(user can choose anyone of them ). How can i start a dtale thread after letting the user choose the dataset. Is there a way around it.
